### DISCUSSION TITLE
Infinite loop

### DISCUSSION THREAD

**Discussion by oswaldofreitas** [vuejs/router#2298] [Help and Questions] [answered]
### Reproduction

https://play.vuejs.org/#eNqVVW1v2zYQ/is3dYBlzBGTLOgHz8nSBcG6YmuLNtiXeUBpibJYS6RAUo4Dw/99dyQl20karEACS+S9PM/dc6dt0nCpsq82mSayabVxsIXcCO7Em7aFHZRGNzBad2I0V9HA6M4JE28yFl731+TX3/G2zYLvXA1RU/wfzxVA1lmRBvfw3uhOuXT0Ct1G42SSRHfENnOiaWv0vyLDWet/8ME6o9Xy6qYzRqgIDVruqumMxTvYbuFHf5GVXV1/xEvY7XwYFuLMFF/3AT95OH9KtQKnL+cJmydXb3UjZmx/801bvsAT5oR16PWGXp5zk/EB4K6SFmryR/DCWJAK/0qpJLKotW5hIXKOVQJdgquImbWiwB/DG+CqQKOlzMktlBEbGZOwkGXGem4zavQx8r+luAcW7eL1DOHHSmMDhqhP5fGXaLR5eCutw59JPAxxD2VzMshjEAiVc1AIvUSJ9Pqhug33/m2vIa1s7LKFS/iHoG9Dv2HERghDYxCF1ZyGNLvJsUno0NQX8OzXY4eQmDz+fZTLYK5DgumWomLziPv0uYKkY584IMXHHQp8rmI5F6LURtzyvEpTpyee6wSU2LgxXKJeyZOy61pk2OB0tHdAxE4POh4DY3COMpAkH547qVWQCuqq0noFttJdXUDF18IL6Etg/gUMol2LYgIL5CwdFFpYNXKEEkCWkP6AebyxDT9n4wAMSQnXGeXxpuNf6Aynqcfs8PzeVwlr9u7zh/fkjWPuH3EipVrK8gFpj8mVvApRC7Tu3Y6TRhufawtZlvVmE4SBMs3FFJzpsNEULlRZbLyOClHyru47iGLupfZ0nVTnccbxwR9gURyW8jqvZb7CyV7qO+3lgXP9u8YWBLHMWDAMTv1c35AT8gxtiKHutVlZ4DjibmRBaednmay4eggDbSkuNckjDlNJMQ9GkujNbG5k68BiG2h/DXOJe+Ll+Xuk6ME+9WUrOxUENJBNY8ujbNvOVmkcoRG6YNdxyXowtCuGSX22vrFefYFbI65wL4e1HNnvdriQ8fz7Gf8Pwgd8ie4h8BDupOEtLjqtEL5nPY8Xdp5Me+nPk30GOp4nlXOtnTLWqXa1zHCbsL3F9UV2wQpcCAdnmbDNycLoexu2deL3BEW+RiNWiLXTurYnvJXfSvDE8Pp19jo7Y7VcMIzOpCrExseOs7lDks5iMUq5fESR1p+shfnQUuuPqfK61vfv/BmN2AA0r0S+eub8q90EyB+NQHZrcUDOcbMUOD10ffv5PY7zwWWji65G6xcuPwnchh1hDGa/dapA2Ad2Hu0fvmM4VXf2duOEsj0pAuqr4e19G29eoL6H+3N2cVBF6x5qYbPc0gcRv6wToM9m8FtoUwgzhfN2AwhWFvDq9PTU78cGw0l1stC4CpopnJ22G3/e8qJAsMOJ36MYFq6Aw0/47wNH71qU+JHqLZPdfwehU8w=

### Steps to reproduce the bug

1. Click on `Go to About` button
2. It shows the About view (works)
3. Go back to `Home`
4. Click on `About` nav link (doesn't work)

### Expected behavior

I expected that when clicking on `About` link it should show the About view

### Actual behavior

Since there is a `param1` passed to the route from the `About` link, when it enters in the `beforeEach` hook in router.js it tries to remove the param in a `newRoute` object and pass to the next() with replace: true, however next when it enters again in the hook I see that the param was not removed as I understand it should.

### Additional information

I had a similar code in my app that always worked, but I noticed that the issue started to happen on 4.3.0, however in the playground I can't make it work even in earlier versions.

**Comment by oswaldofreitas** [comment]
IDK if it's related, but I created [another issue](https://github.com/vuejs/router/issues/2299) (which actually represents better the problem in my project)

**Comment by posva** [comment, marked as answer]
In this case, it's a bit different. The manipulation of params is incorrect. [Here is a fixed version](https://play.vuejs.org/#eNqVVetu2zYUfpUzdYBkzBGTLOgPz8nSBcG6YZeiC/ZnHlBaoi3WEimQlOPA0LvvHJKSnUuDFYkh8dz4feemfdJwqfLPNpklsmm1cbCHwgjuxLu2hR5WRjeQbjuRLlQ0MLpzwkRNzsLxoCa/QcfbNg++CzVGzfA3WSiAvLMiC+7h3OhOuSx9g27pJJkm0R2xzZ1o2hr9r8hw3voHvlhntFpf3XTGCBWhQctdNZuzqIP9Hr71inzV1fUHVELf+zAsxJkrvh0CfvRwfpNqA05fLhK2SK7e60bM2UHzRVu+RAlzwjr0ekeHl9xkfAG4q6SFmvwRvDAWpML/lVQSWdRat7AUBccsgV6Bq4iZtaLEh+ENcFWi0VoW5BbSiIWMl7Bwy5wN3OZU6MfI/5biHli0i+o5wo+ZxgKMUZ+3x++i0ebhvbQOH9MoDHGP2+ZkbI+xQSidY4fQIbbI0D+Ut1HvT4ce0srGKlu4hH8I+j7UG1KWIgyNQRRmcxau6afBRPFGoIkvEJoNHqFgM5/Psx8f+wccFODfJ1cbvPqYb7anS7CWlIrZS/nJJh5HAI6vPfb7QsXsLsVKG3HLiyrLnJ566hO4xMYlH7pX1yLHSmfpwRSxOj029AQYg3PsB0l9xAsntQo9gw1Wab0BW+muLqHiW+E76VPg/AkM4tyKcgpLZCsdlFpYlTrCx1gplt16LQwBkSvIvsE7vaMNj7NJAInUhOuMAmc6QQKcL/ob5dke8jwndsF9Bnt/jtGi+AxrkvZTdMIOLLBgFA76yQ9DxsTOt0gpVryrh2pgnw5d9HxTVOdxfPHFC5Cmw+RcF7UsNji0a32nfalxZH/WmNRQ+DkLhsFpGNkbcpJqHRIbQ91rs7HAcXpdakFp58eUrLh6iHwpLqXdIw4DRzGPpo3ozW1hZOvAYtJoNY0jhyvg9dF60p2jfebTtupUaImRbBYLF1uw7WyVxXFI0QXLh/vTg6E1MA7hi/mN+RoS3BpxhSs3bNzIvu9x16L86xn/D8JHfInuMfAQ7qThLe4wrRC+Z72ICrtIsBFDaRfJ4QYSL5LKudbOGOtUu1nnuBnYweL6Ir9gJQ73kSwXtjlZGn1vwyJO/MxT5Gs0wlnaOq1re8Jb+aULnhlev83f5meslkuG0ZlUpdj52HHIeiTpLCZjJddPKNIqk7Uwf7ZU+sdUeV3r+1+9jEZsBFpUoti8IP9sdwHyByOQ3VYckXPcrAVOD6lv//pD7PB9VDa67Gq0fkX5UeB+6whjMPupUyXCPrLzaH/xFcOpurO3OyeUHUiNK8fb+zLevEL9APf7/OIoi9Y91MLmhaVvHX40p0BfxOC31KYUZgbn7Q4QrCzhzenpKW4lQCOzlupkqXEVNDM4O213Xt7yskSwo8QvRAwLV8DhO/z5wNG7Fiv84AyWSf8fz8FFhw==).

- You need to use route names in order to be able to change a route based on the params
- Spread the whole route instead of using json copy. Some properties are intentionally not enumerable to avoid being copying when using regular methods like spread, assign, Object.keys, etc
